Use temp stream instead of string to buffer content

yookoala opened the pull request:
* Update MultipartStreamBuilder::build to buffer built content
  in a php://temp stream resource instead of (string) $contents.
  This way, the size of the stream built won't be limited by PHP
  runtime limit.

| Q               | A
| --------------- | ---
| Bug fix?         | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets |
| Documentation   | Fully backward compatible without any interface changes. Only internal programming changes.
| License         | MIT


#### What's in this PR?

Rewrite to append stream contents and boundaries into a ["php://temp"](https://www.php.net/manual/en/wrappers.php.php) stream as a buffer. Contents within predefined limit (2MB) will be handled in memory for speed. Otherwise, PHP will create a temp file for storage. This mean the multipart stream size is virtually without size limit (unless if, of course, your harddisk is full).


#### Why?

The original MultipartStreamBuilder::build would read everything into a string variable `$streams` before making it another stream. If the total size of the content exceed the PHP's memory limit, the process would fail.

#### Checklist

- [x] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [x] Documentation pull request created (if not simply a bugfix)

yookoala committed:
Use temp stream instead of string to buffer content

* Update MultipartStreamBuilder::build to buffer built content
  in a php://temp stream resource instead of (string) $contents.
  This way, the size of the stream built won't be limited by PHP
  runtime limit.

dbu reviewed (approved):
looks good to me. i can imagine that this adds some filesystem overhead for medium sized streams where previously things worked in memory and now go to the temporary file, but if people really need to worry on that level they can configure the temp handling, according to php.net

 @GrahamCampbell do you agree or are there any doubts?

yookoala committed:
Detect available memory for maxmemory in buffer

* Detect available memory (in bytes). Then config the buffer
  stream to stay in memory until reaching 1/4 of that.
* Added setBufferMaxMemory to manually override that limit.

yookoala commented:
Updated by PR to include mechanism to detect available memory and assign maxmemory to the buffer (size of buffer suitable to stay in memory).

It would subtract from [memory_get_usage()](https://www.php.net/manual/en/function.memory-get-usage.php) result from PHP `memory_limit` to get the available memory. Then use 1/4 as the maxmemory setup of the ["php://temp"](https://www.php.net/manual/en/wrappers.php.php) stream used for buffering.

I also added a setBufferMaxMemory() public method for overriding this.

dbu commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +373,50 @@ private function createStream($resource)
+    /**
+     * Get and parse memory_limit into bytes integer.
+     *
+     * @return integer
```
```suggestion
```

no need for this, the return type declaration already has this information.

dbu commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +373,50 @@ private function createStream($resource)
+     *     Size of stream data buffered (in bytes) until using temporary
+     *     file to buffer.
+     *
+     * @return MultipartStreamBuilder
```
```suggestion
```

same here, the return type declaration is enough

dbu commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +373,50 @@ private function createStream($resource)
+     * in memory if the size of the stream is smaller than this size.
+     * Otherwise, PHP will store the stream data in a temporary file.
+     *
+     * @param integer $size
```
```suggestion
     * @param int $size
```

dbu reviewed (commented):
this is really cool! i think using max 1/4 of available memory by default is a good idea to not use up all the memory with large streams.

yookoala committed:
Fix code style

yookoala commented:
Finally fixed all the style issues :-)

yookoala committed:
Improve documentation to getAvailableMemory

* Improve documentation to MultipartStreamBuilder::getAvailableMemory

dbu reviewed (approved):
cheers, looks good to me. i wait until tomorrow so @Nyholm or @GrahamCampbell get a chance to see it before merging.

@yookoala  could you please add a changelog note for this improvement?

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -133,28 +138,52 @@ public function addResource($name, $resource, array $options = [])
-            // Convert the stream to string
-            /* @var $contentStream StreamInterface */
+            /**
+             * @var \Psr\Http\Message\StreamInterface
```
should be single line (but with double star)

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+     */
+    protected static function getAvailableMemory(): int
+    {
+        $memory_limit = ini_get('memory_limit');
```
This does not get the memory limit, it just gets the limit written in the config. Also, that function might be disabled, which would trigger an error.

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+            throw new \Exception("Unknown memory_limit format: {$memory_limit}");
+        }
+        if ('G' === $matches[2]) {
+            $memory_limit = $matches[1] * 1024 * 1024 * 1024; // nnnG -> nnn GB
```
$memoryLimit

Nyholm commented on src/MultipartStreamBuilder.php:
```diff
@@ -133,28 +138,52 @@ public function addResource($name, $resource, array $options = [])
+        // Open a temporary read-write stream as buffer.
+        // If the size is less than predefined limit, things will stay in memory.
+        // If the size is more than that, things will be stored in temp file.
+        $buffer = fopen('php://temp/maxmemory:'.$maxmemory, 'r+');
```
Lets use `php://temp` instead. Ie, skip the max memory. That will automatically switch to file after a few mega bytes. 

It is better than adding a bunch of logic here.

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+        $memory_limit = ini_get('memory_limit');
+        if ('-1' === $memory_limit) {
+            // If there is no memory limit, return 100MB by default.
+            return 100 * 1024 * 1024;
```
Bad idea. We should return 128MB (the default) minus the used memory.

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+     * and memory_get_usage(). If memory_limit is "-1", the default estimation
+     * would be 100MB.
+     *
+     * @throws \Exception if the ini format does not match expectation
```
Better to return the default than to crash.

Nyholm commented on src/MultipartStreamBuilder.php:
```diff
@@ -133,28 +138,52 @@ public function addResource($name, $resource, array $options = [])
+            if ($contentStream->isReadable()) {
+                while (!$contentStream->eof()) {
+                    // read 8KB chunk into buffer until reached EOF.
+                    fwrite($buffer, $contentStream->read(8192));
```
For a large steam, this will consume 8MB, lets reduce it to 1MB chunks instead. You will no lose any performance.

yookoala replied:
This would read 8KB a chunk. I can update it to 1MB.

Nyholm replied:
Hehe, Math. =)

Yes please.

GrahamCampbell commented:
I've left some comments. There are some issues, some of which are major (but easy to resolve).

Nyholm commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+     * @param int $size size of stream data buffered (in bytes)
+     *                  until using temporary file to buffer
+     */
+    public function setBufferMaxMemory(int $size): MultipartStreamBuilder
```
This can be removed

yookoala committed:
Update CHANGELOG.md

Nyholm commented on src/MultipartStreamBuilder.php:
```diff
@@ -342,4 +371,47 @@ private function createStream($resource)
+     *
+     * @throws \Exception if the ini format does not match expectation
+     */
+    protected static function getAvailableMemory(): int
```
This can also be removed.

GrahamCampbell replied:
👍

Nyholm commented on src/MultipartStreamBuilder.php:
```diff
@@ -38,6 +38,13 @@ class MultipartStreamBuilder
+     *          To have MultipartStreamBuilder manage it automatically, set to -1.
+     *          Default: -1
+     */
+    private $bufferMaxMemory = -1;
```
This can be removed.

Nyholm reviewed (changes requested):
Thank you

I agree that we should use a stream, but lets fallback to PHP default behavior.

GrahamCampbell commented:
@Nyholm has suggested some excellent changes. 👍

yookoala committed:
Remove the memory detection logic from build

* Remove the memory detection logic from MultipartStreamBuilder::build.

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -133,28 +131,44 @@ public function addResource($name, $resource, array $options = [])
+                }
             } else {
-                $streams .= $contentStream->getContents();
+                // Try to getContents for non-readable stream.
```
this doesn't read well

yookoala replied:
Mimicing the behaviour of [Laminas](https://github.com/laminas/laminas-httphandlerrunner/blob/4c952795f1ee141aa4c1d817995d6f76015ea07d/src/Emitter/SapiStreamEmitter.php#L61-L77) (previously Zend framework) here.

yookoala replied:
Or if you're referring to the comment line, please advice how to rewrite it.

GrahamCampbell replied:
I was referring to the line I commented on. The comment text.

yookoala committed:
Code style update

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -140,9 +140,7 @@ public function build()
-            /**
-             * @var \Psr\Http\Message\StreamInterface
-             */
+            /* @var $contentStream StreamInterface */
```
Needs to have two stars, not one. A single star is a comment, and not phpdoc. If php-cs-fixer is messing this up, you probably need to replace this with `/** @var StreamInterface */` and/or just disable the broken fixer.

yookoala committed:
Minor updates

* Improve readability of comment.
* Use the old __toString approach if the stream is not isReadable().

GrahamCampbell commented on src/MultipartStreamBuilder.php:
```diff
@@ -149,13 +149,13 @@ public function build()
-                // Less controllable chunk size (thus memory usage).
-                fwrite($buffer, $contentStream->getContents());
+                // If not isReadable, read from __toString().
+                // Warning: This could attempt to load a large amount of data into memory.
```
Don't see the point in this warning? Who is the comment intended for? End users will not read this.

yookoala replied:
I, as end user, would read code from time to time. But if it makes you happy, I can remove the warning line.

yookoala replied:
How else, if needed, do you want me to change it?

Nyholm replied:
Yes, lets remove both these comments. 

```
// If not isReadable, read from __toString().
// Warning: This could attempt to load a large amount of data into memory.
```

yookoala committed:
Remove unnecessary comment

dbu reviewed (approved):
looks good to me, thanks!

Nyholm reviewed (approved):
Cool. Im really happy with this now.

GrahamCampbell reviewed (approved):
Good work.
